### PR TITLE
Migrate bernoulli from CUDA_tensor_apply2 to TensorIterator

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36135 Completely kill CUDAApplyUtils.cuh
* **#34006 Migrate bernoulli from CUDA_tensor_apply2 to TensorIterator**
* #34005 Migrate gamma from CUDA_tensor_apply2 to TensorIterator
* #34004 Migrate poisson from CUDA_tensor_apply2 to TensorIterator

The CUDA 9 curand workaround is removed in order to be able to use
TensorIterator. I don't know "have great perf on old CUDA" vs "
be able to completely kill CUDA_tensor_apply2" which is more valuable.

Differential Revision: [D20196995](https://our.internmc.facebook.com/intern/diff/D20196995)

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
On CUDA 10.2, there is a perf drop of 5% due to loss of unrolling of curand

```python
import torch
print(torch.__version__)

# warm up
for i in range(1000):
    torch.arange(100000, device='cuda')

def benchmark(size):
    s1 = 2 ** (size // 2)
    s2 = 2 ** (size - size // 2)
    p = torch.rand((s1, s2), device='cuda')
    d = torch.distributions.bernoulli.Bernoulli(p)
    torch.cuda.synchronize()
    %timeit d.sample(); torch.cuda.synchronize()

for i in range(20, 31):
    benchmark(i)
```

Before:
```
1.5.0a0+9956a23
30.8 µs ± 27.7 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
46.5 µs ± 17.9 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
78 µs ± 122 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
141 µs ± 1.01 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
266 µs ± 41.1 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
520 µs ± 5.54 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.09 ms ± 52.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
2.13 ms ± 119 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
4.33 ms ± 170 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
8.84 ms ± 931 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
16.1 ms ± 244 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

After:
```
1.5.0a0+77c9c3c
30.7 µs ± 661 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
48.5 µs ± 407 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
84.3 µs ± 1.42 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
158 µs ± 3.51 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
295 µs ± 1.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
575 µs ± 2.04 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.14 ms ± 1.82 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
2.29 ms ± 22.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
4.63 ms ± 21.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
9.23 ms ± 201 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
16.8 ms ± 77.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
